### PR TITLE
Update tc-print to work on Aarch64.

### PR DESCRIPTION
**Pull request by swalk-cavium**
These changes update the tc-print tool on Aarch64 hosts  so that it can print out the files generated with -vEval.DumpTC=true.  This reuses disassembler in the vixl module with a
slight change.  During Initial implementation of the Aarch64 host the capability of changing
the base address of the disassembler was removed from vixl.  This change adds back in part
of that.

This change also adds an option, -o, so that the binary opcodes are emitted in the report.
I find this useful in finding sequences to improve.  The default is off as before.

I do not have access to a powerpc64 host, but the disassembly code should be very similar.

These changes compile and run on both Aarch64 and x86_64 hosts.

### CODE REVIEW

**Comment by jim-saxman** [comment]
@swalk-cavium Hi, is there any chance you could rebase this request?

**Comment by swalk-cavium** [comment]
@jim-saxman - I could, but it's not clear that it will resolve the merge.  I changed the
name of one of the files since it wasn't x86 specific anymore.  If you look at the merge
complain it's down in an #ifdef

**Comment by swalk-cavium** [comment]
@jim-saxman - I think I misinterpreted the git complaint.  You should be able to build
the executable now.  I have not validated the interaction with the oss-performance perf
scripts yet.  It appears that information was updated relatively recently.

**Comment by jim-saxman** [comment]
@swalk-cavium success! Thanks. It works well with OSS, following the README.md instructions.

**Comment by swalk-cavium** [comment]
@jim-saxman - The -o option prints out the opcodes.  I find it useful to figure out where the pc-relative branches are going.  It will also try to resolve symbols (same as before) if the /tmp/hhvm.nm file exists.  The parsing was dodgy, but this works for me

nm --demangle ../hhvm/hphp/hhvm/hhvm | awk '{print $1,$2,$4}' >/tmp/hhvm.nm
../hhvm/hphp/tools/tc-print/tc-print -o 

    0x3e400238: f2b55ec1                                                          movk x1, #0xaaf6, lsl #16
    0x3e40023c: f2dfffe1                                                          movk x1, #0xffff, lsl #32
    0x3e400240: d2800022                                                          movz x2, #0x1
    0x3e400244: d29b9512                                                          movz x18, #0xdca8
    0x3e400248: f2a02152                                                          movk x18, #0x10a, lsl #16
    0x3e40024c: d63f0240                                                          blr x18  # HPHP::Func . //<<---
    0x3e400250: 58000052                                                          ldr x18, pc+8 (addr 0x3e400258)
    0x3e400254: d61f0240                                                          br x18

**Comment by jim-saxman** [comment]
jsaxman@art-ral-aw1:/home/jsaxman/oss-performance$ grep "0x" mw_cbz_noDead_top50.tcprint|wc -l
8907
jsaxman@art-ral-aw1:/home/jsaxman/oss-performance$ grep "0x" mw_cbz_noDead_top50.tcprint |grep "b\."|wc -l
706
jsaxman@art-ral-aw1:/home/jsaxman/oss-performance$ grep "0x" mw_cbz_noDead_top50.tcprint |grep "blr "|wc -l
307
jsaxman@art-ral-aw1:/home/jsaxman/oss-performance$ grep "0x" mw_cbz_noDead_top50.tcprint |grep "b #"|wc -l
730
jsaxman@art-ral-aw1:/home/jsaxman/oss-performance$ grep "0x" mw_cbz_noDead_top50.tcprint |grep "unallocated"|wc -l
816
jsaxman@art-ral-aw1:/home/jsaxman/oss-performance$ grep "0x" mw_cbz_noDead_top50.tcprint |grep "tst"|wc -l
163
jsaxman@art-ral-aw1:/home/jsaxman/oss-performance$ grep "0x" mw_cbz_noDead_top50.tcprint |grep "tbz"|wc -l
0
jsaxman@art-ral-aw1:/home/jsaxman/oss-performance$ grep "0x" mw_cbz_noDead_top50.tcprint |grep "tbnz"|wc -l
0
jsaxman@art-ral-aw1:/home/jsaxman/oss-performance$ grep "0x" mw_cbz_noDead_top50.tcprint |grep "cmp"|wc -l
434
jsaxman@art-ral-aw1:/home/jsaxman/oss-performance$ grep "0x" mw_cbz_noDead_top50.tcprint |grep "cbz"|wc -l
0
jsaxman@art-ral-aw1:/home/jsaxman/oss-performance$ grep "0x" mw_cbz_noDead_top50.tcprint |grep "str"|wc -l
422
jsaxman@art-ral-aw1:/home/jsaxman/oss-performance$ grep "0x" mw_cbz_noDead_top50.tcprint |grep "stur"|wc -l
503
jsaxman@art-ral-aw1:/home/jsaxman/oss-performance$ grep "0x" mw_cbz_noDead_top50.tcprint |grep "ldur"|wc -l
179
jsaxman@art-ral-aw1:/home/jsaxman/oss-performance$ grep "0x" mw_cbz_noDead_top50.tcprint |grep "ldr"|wc -l
1310
jsaxman@art-ral-aw1:/home/jsaxman/oss-performance$ grep "0x" mw_cbz_noDead_top50.tcprint |grep "mov"|wc -l
1462
jsaxman@art-ral-aw1:/home/jsaxman/oss-performance$ grep "0x" mw_cbz_noDead_top50.tcprint |grep "movz"|wc -l
715
jsaxman@art-ral-aw1:/home/jsaxman/oss-performance$ grep "0x" mw_cbz_noDead_top50.tcprint |grep "movk"|wc -l
437
jsaxman@art-ral-aw1:/home/jsaxman/oss-performance$ grep "0x" mw_cbz_noDead_top50.tcprint |grep "mul"|wc -l
0
jsaxman@art-ral-aw1:/home/jsaxman/oss-performance$ grep "0x" mw_cbz_noDead_top50.tcprint |grep "div"|wc -l
0
jsaxman@art-ral-aw1:/home/jsaxman/oss-performance$ grep "0x" mw_cbz_noDead_top50.tcprint |grep "lsl"|wc -l
469
jsaxman@art-ral-aw1:/home/jsaxman/oss-performance$ grep "0x" mw_cbz_noDead_top50.tcprint |grep "lsr"|wc -l
13
jsaxman@art-ral-aw1:/home/jsaxman/oss-performance$ grep "0x" mw_cbz_noDead_top50.tcprint |grep "add"|wc -l
861
jsaxman@art-ral-aw1:/home/jsaxman/oss-performance$ grep "0x" mw_cbz_noDead_top50.tcprint |grep "sub"|wc -l
228
jsaxman@art-ral-aw1:/home/jsaxman/oss-performance$ grep "0x" mw_cbz_noDead_top50.tcprint |grep "crc"|wc -l
0
jsaxman@art-ral-aw1:/home/jsaxman/oss-performance$ grep "0x" mw_cbz_noDead_top50.tcprint |grep "nop"|wc -l
62
jsaxman@art-ral-aw1:/home/jsaxman/oss-performance$ grep "0x" mw_cbz_noDead_top50.tcprint |grep "brk"|wc -l
0
jsaxman@art-ral-aw1:/home/jsaxman/oss-performance$ grep "0x" mw_cbz_noDead_top50.tcprint |grep "tst"|wc -l
163
jsaxman@art-ral-aw1:/home/jsaxman/oss-performance$ grep "0x" mw_cbz_noDead_top50.tcprint |grep "ldp"|wc -l
183
jsaxman@art-ral-aw1:/home/jsaxman/oss-performance$ grep "0x" mw_cbz_noDead_top50.tcprint |grep "stp"|wc -l
12

**Comment by jim-saxman** [comment]
sorry for the above ugliness, I'm in a rush right now.

**Comment by swalk-cavium** [comment]
@jim-saxman - What are you trying to show?  Some kind of instruction use distribution?

**Comment by jim-saxman** [comment]
sorry, That was meant for an internal JIRA.

**Comment by swalk-cavium** [comment]
@mxw - Hi Max,  Any comments on this one?

File: hphp/tools/tc-print/offline-code.cpp [hunk 1]
```diff
@@ -124,7 +121,326 @@ TCA OfflineX86Code::getTransJmpTargets(const TransRec *transRec,
...
+  TCRegion tcr = findTCRegionContaining(startAddr);
+  disasm(tcRegions[tcr].file, tcRegions[tcr].baseAddr, startAddr, len,
+         perfEvents, BCMappingInfo(tcr, bcMap), true, hostOpcodes);
+}
+
+void OfflineCode::loadSymbolsMap() {
+  loadSymbolsMapNm();
+}
+
+void OfflineCode::loadSymbolsMapNm() {
```
Comment (ottoni) [inline hphp/tools/tc-print/offline-code.cpp]: Hmm, I folded this into the function above a couple months back in 83a65510caa5deea59c59deb69684faaee474d3c. Looks like your rebase lost that, so please fix it.

File: hphp/tools/tc-print/tc-print.cpp [hunk 2]
```diff
@@ -279,6 +280,9 @@ void parseOptions(int argc, char *argv[]) {
         if (optopt == 'd' || optopt == 'c' || optopt == 'p' || optopt == 't') {
           fprintf (stderr, "Error: -%c expects an argument\n\n", optopt);
         }
+      case 'o':
```
Comment (ottoni) [inline hphp/tools/tc-print/tc-print.cpp]: Please document this new option in the usage() function in tc-print.cpp.

File: hphp/tools/tc-print/offline-code.cpp [hunk 3]
```diff
@@ -124,7 +121,326 @@ TCA OfflineX86Code::getTransJmpTargets(const TransRec *transRec,
...
+    frontier += kInstructionSize;	
+    ip += kInstructionSize;
+  }
+}
+
+#elif defined(__x86_64__)
+#if defined(HAVE_LIBXED)
+
+// x86 below here
+// TODO: move to separate file
```
Comment (ottoni) [inline hphp/tools/tc-print/offline-code.cpp]: I agree that putting the target-specific code on separate files would be better. Could you please do so?

**Comment by swalk-cavium** [comment]
@ottoni - I keep forgetting to add the reminder.  I think I've made the changes as you requested.
It compiles and runs.  Can you review?  Thanks.

**Comment by swalk-cavium** [comment]
@mxw - Hi Max,  Do I have to do anything else for this one?

**Comment by mxw** [comment]
@swalk-cavium—Nope, sorry; I'll take care of landing.

**Pull request closed without merging**